alpha/lib/model/ServerNodePeer.php

```diff
@@ -134,13 +134,17 @@ public static function getCacheInvalidationKeys()
 		return array(array("serverNode:id%s", self::ID), array("serverNode:hostName=%s", self::HOST_NAME));
 	}
 
-	public static function retrieveActiveUnoccupiedServerNodesByType($type, PropelPDO $con = null)
+	public static function retrieveActiveUnoccupiedServerNodesByType($type, $tag, PropelPDO $con = null)
 	{
 		$c = new Criteria();
 		$c->add(ServerNodePeer::STATUS, ServerNodeStatus::ACTIVE);
 		$c->add(ServerNodePeer::TYPE, $type);
 		$c->add(ServerNodePeer::HEARTBEAT_TIME, time() - ServerNode::SERVER_NODE_TTL_TIME, Criteria::GREATER_EQUAL);
 		$c->addOr(ServerNodePeer::HEARTBEAT_TIME, null);
+		if (!empty($tag))
+		{
+			$c->add(ServerNodePeer::TAGS, $tag, Criteria::LIKE);
```

Comment:
Using like query on string field which runs on the mysql side is not that good. I think it would be better to fetch all and filter by tag.